pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pod
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+)
+
+const (
+	ScheduleingGateName = "kueue.x-k8s.io/scheduling-gate"
+	FrameworkName       = "pod"
+
+	gateNotFound = -1
+)
+
+var (
+	gvk = corev1.SchemeGroupVersion.WithKind("Pod")
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:  SetupIndexes,
+		NewReconciler: NewReconciler,
+		SetupWebhook:  SetupWebhook,
+		JobType:       &corev1.Pod{},
+	}))
+}
+
+// +kubebuilder:rbac:groups=scheduling.k8s.io,resources=priorityclasses,verbs=list;get;watch
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update;patch
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=pods/status,verbs=get
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/finalizers,verbs=update
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors,verbs=get;list;watch
+
+var NewReconciler = jobframework.NewGenericReconciler(
+	func() jobframework.GenericJob {
+		return &Pod{}
+	}, nil)
+
+type Pod corev1.Pod
+
+var _ jobframework.GenericJob = (*Pod)(nil)
+var _ jobframework.JobWithCustomStop = (*Pod)(nil)
+
+// Object returns the job instance.
+func (j *Pod) Object() client.Object {
+	return (*corev1.Pod)(j)
+}
+
+func (p *Pod) gateIndex() int {
+
+	for i := range p.Spec.SchedulingGates {
+		if p.Spec.SchedulingGates[i].Name == ScheduleingGateName {
+			return i
+		}
+	}
+	return gateNotFound
+}
+
+// IsSuspended returns whether the job is suspended or not.
+func (p *Pod) IsSuspended() bool {
+	return p.gateIndex() != gateNotFound
+}
+
+// Suspend will suspend the job.
+func (p *Pod) Suspend() {
+	// TODO: maybe change the framework so this can provide feedback,
```

Comment:
This will have to look almost exactly as preemptions by kube-scheduler:

We add a `DisruptionTarget` condition and call `Delete`, which adds a `deletionTimestamp`.